### ISSUE TITLE
Scikit-Tree: A Python package for modern decision tree models

### ISSUE THREAD

**Issue by adam2392** [scikit-learn-contrib/scikit-learn-contrib#61]
Request for project inclusion in scikit-learn-contrib

* Project name: scikit-tree
* Project description: Scikit-learn compatible decision trees
* Authors: Adam Li and members of NeuroData Lab
* Current repository: https://github.com/neurodata/scikit-tree
* Requirements:
 * [x] scikit-learn compatible (check_estimator passed)
 * [x] Documentation (guide, API reference, example gallery)
 * [x] Unit tests (coverage: 91%)
 * [x] Python3 compatible
 * [x] PEP8 compliant
 * [x] Continuous integration

As of 04/11/23, we are not 100% ready to transfer this over yet, but I wanted to get the process started to make sure our design approach is fine for including into scikit-learn-contrib.

cc: @jovo

**Comment by adrinjalali** [comment]
I think this is great, but here:

> Dependency on a fork of scikit-learn Due to the current state of scikit-learn's internal Cython code for trees, we have to instead leverage a maintained fork of scikit-learn at https://github.com/neurodata/scikit-learn, where specifically, the fork branch is used to build and install this repo. We keep that fork well-maintained and up-to-date with respect to the main sklearn repo. The only difference is the refactoring of the tree/ submodule.

I would suggest instead of forking the whole repo, to vendor the parts of scikit-learn you need to change inside `scikit-tree`, and build with the project. This way you'd depend on the upstream `scikit-learn`, and yet able to do the modifications you need on the tree code.

Also note that the tree code is going under substantial changes. cc'ing @thomasjpfan here.

**Comment by adam2392** [comment]
Hi @adrinjalali thanks for looking thru!

Slightly a long answer because a lot of thought and playing around went into this, so wanna make sure I have everything documented in one post:

We prefer a repo-fork because it is a lot easier to sync upstream changes of sklearn to a fork than it is to sync changes to a copy/pasted tree code. I chose this approach after seeing how far quantile trees and econml has diverged from sklearn where it might as well be two different tree code bases. We do not want this for a few reasons:

1. because it makes inclusion of oblique trees and sklearn-includable generic tree models (e.g. causal trees w/ 1000s of citations, quantile trees, etc.) more difficult. 
2. This design pattern forces scikit-tree to be in line with upstream essentially. If a PR of oblique trees is opened at any time, it is guaranteed to be pseudo-compatible. 
3. It also demonstrates the utility of refactoring proposed in upstream sklearn. My longer-term goal is to help sklearn review Tree code and Cython code to alleviate maintainer burden and if they are willing in the future, help contribute refactorings that make reliance on the fork unnecessary. All this w/o loss of performance ideally.
4. With this design, we can demonstrate implementation of quantile trees (which I understand sklearn is interested in potentially taking since it is not maintained anymore) in prolly < 100 LOC. It is a Cython class that overrides how the leaf nodes are set and then that is it.
5. With this design, we can also demonstrate implementation of causal trees relatively easily.

By adding new tree models, previous packages (e.g. econml and zillow/quantiletree) essentially vendor the tree submodule to do what is necessary. However, just a simple example of compiling against Cython3.0. Both packages will now raise a boatload of warnings, whereas this has been fixed in upstream sklearn already. In order to maintain this  change, or be aware, they would have to do a side-by-side comparison (w/o GitHub) to understand changes they need to make to their vendored tree submodule vs the new tree code they added.

##### Relevant tradeoffs of fork vs vendored tree copy
The changes I made to the fork are non trivial, but they do not touch the existing tree code a lot. Therefore even changes like missing value support is easily synced in with a push of a button and minor refactoring. Then scikit-tree has a minor job refactoring. Even when tree submodule is touched, the upstream changes are much easier to include this way. The main con is if for some reason the tree submodule gets overhauled very quickly where it is difficult to parse how to pass changes downstream to the sklearn-fork and scikit-tree. However, based on my convos w/ sklearn devs, that is not possible.

Otoh, if I wanted to support changes from upstream sklearn in a vendored tree submodule, it would not have nice features like git sync and have to go thru essentially the same if not more refactorings. 

Our strategy basically splits maintenance into two parts to therefore make life easier: i) keep fork up to date and refactor only if necessary and ii) refactor scikit-tree if any tree upstream code changes.

Wdyt?

**Comment by adam2392** [comment]
> I understand the logic from a maintenance / dev point of view. But from a user's point of view, it's a terrible design because `scikit-learn` and `scikit-learn-tree` are two different packages on `pypi`. Libraries have dependencies to `scikit-learn`, which means you can't install anything along with `scikit-tree` which depends on `scikit-learn`, otherwise you have two packages exposing the same namespace.
> 
> I see two paths forward if you want to include this project in the ecosystem:
> 
> * try to get your changes to tree to upstream, and then depend on that from `scikit-tree`
> * vendor the tree code, and doing the git there is not that difficult, one doesn't need GitHub to do git, there are enough alternatives out there.

**Comment by adrinjalali** [comment]
> Re alternative: Do you have any suggestions on a workflow that makes maintenance easy? If we copy/paste the tree submodule and apply the modifications that work for us, then how do we sync changes that occur in scikit-learn upstream w/o line-by-line manually comparing?

I was thinking submodules, but wasn't sure, so I asked chatGPT, and I think it's on the right track:

---------------

One common way to vendor a file from another repository and keep it up to date with upstream changes is to use Git submodules. Submodules allow you to include a repository as a subdirectory within another repository, and keep track of its commit history separately.

Here's an example workflow:

1. Add the external repository as a submodule to your repository using the `git submodule add` command. For example, if you want to vendor a file from the "external-repo" repository, you can run:

```bash
git submodule add https://github.com/external-repo.git vendor/external-repo
```


This will add the "external-repo" repository as a submodule in a new directory called "vendor/external-repo" in your repository.

2. Update the submodule to the latest upstream version by running `git submodule update --remote`. This will fetch the latest changes from the upstream repository and update the submodule to the latest commit.

3. Make changes to the submodule as needed. If you make changes to the submodule, you can commit them in the usual way within the submodule directory.

4. Update the submodule reference in the parent repository. If you make changes to the submodule, you need to update the submodule reference in the parent repository by committing the changes to the submodule reference. You can do this by running `git add vendor/external-repo` and then committing the change.

5. Pull upstream changes to your repository. When upstream changes are made to the submodule, you can pull them in by running `git submodule update --remote` again, which will update the submodule to the latest upstream commit. Then commit the submodule reference update in the parent repository.

By following this workflow, you can keep the file from the external repository up to date with upstream changes while keeping it separate from your own codebase.

**Comment by adam2392** [comment]
Hmm I will have to look into this. Thanks for the tip. I wasn't even thinking of this strategy, but it makes sense.

So the strategy would be to create a submodule of our scikit-learn fork inside scikit-tree, so that way we don't have to create an explicit dependency on it and then depend on main scikit-learn for non-tree stuff (e.g. importing BaseEstimator, or using any other things like metrics). Do I understand this correctly?

Then we can maintain both separately as is, but remove this dependency issue.

The only open issue I'm not sure of (or even how to prompt ChatGPT for) is how to handle building... I use meson-python building since that's what's done w/ scipy, but scikit-learn uses setuptools. Any idea how to handle this nicely?

**Comment by adam2392** [comment]
I believe I resolved the issue where there are now two separate namespaces: `sklearn` and `sklearn_fork`. The `sklearn_fork` contains any stuff that needs to be imported due to `tree/` submodule changes.

**Comment by adam2392** [comment]
The issue is that scikit-learn uses setuptools to build and I went with a meson build system. So this is basically not possible unless I rewrite a setuptools build. 

Do you think it's time to move scikit learn to meson since numpy scipy and others have as well?

**Comment by adrinjalali** [comment]
another option is not to build the whole scikit-learn, and only treat that git submodule as a part of your own package, and include the files you want to compile in your meson build.

But I think we'd like to move to meson at some point anyway.

**Comment by adam2392** [comment]
Hmm I suppose that could be a soln. I'll have to look into it.

I think that's pretty difficult since the `tree` submodule inherently depends on possibly other modules that require some Cython. Minimally, e.g. `sklearn/utils/` needs to get compiled. Idk if more does...

**Comment by adam2392** [comment]
I've refactored the code to leverage `neurodata/scikit-learn:submodule` as a submodule inside `sktree/_lib/`, which then gets used for the changes made inside the `tree` and `ensemble` modules of sklearn. Otherwise, the regular sklearn from main branch is used.

As a pip dependency tho, only sklearn is required. 

When building and releasing sktree, we will automatically pull in the submodule to leverage certain Cython files and changes.

Lmk if this looks good @adrinjalali ?

**Comment by adrinjalali** [comment]
that sounds great, I noticed this is not on `pypi` though, could you have a release and have the installation be an easy `pip install scikit-tree`?

**Comment by adam2392** [comment]
Sounds good! I realized that we currently depend on scikit-learn v1.3 private API changes, so I'm going to wait until that gets released, so I don't have to build from source sklearn:main. I'll ping back here when that's done.

**Comment by adam2392** [comment]
Okay, I have slated a release of v0.1 at https://pypi.org/project/sktree/. I think I named it `sktree` instead of `scikit-tree`. I can fix it to be `scikit-tree` when we do the next release. The wheels should also be distributed for combos of Python versions and common OSes. Lmk if the install works @adrinjalali ?

**Comment by adrinjalali** [comment]
The package doesn't have a dependency set for it, so installing it doesn't pull the right other packages for it to actually work.

**Comment by adam2392** [comment]
Ah yes I did not add `dependency` block in the `pyproject.toml` file. It should work now. https://pypi.org/project/scikit-tree/

**Comment by adrinjalali** [comment]
Seems to work, but your CI seems to be failing, and your README still refers to a fork of scikit-learn.

**Comment by adam2392** [comment]
> Seems to work, but your CI seems to be failing, 

Hmm okay some weird windows issue. I will look into it.

> and your README still refers to a fork of scikit-learn.

This snippet is just conveying some changes that were made in our submodule dependency so users are aware this change occurs. Where should it go instead?

**Comment by adam2392** [comment]
I updated the README to be a bit more clear on how the submodule affects things, so hopefully that addresses the issue now.

**Comment by jovo** [comment]
@adam2392 Is the CI still failing? If not, is there anything else required by you to get this complete?

**Comment by adam2392** [comment]
No. This is all good to go from our end.

@adrinjalali if this looks good to you, let us know! Thanks for the patience

**Comment by adam2392** [comment]
@adrinjalali it seems you're on an older version for some reason? Is the pip installing from a cache?

```
(test) adam2392@adams-mbp-5 scikit-tree % pip install scikit-tree
Collecting scikit-tree
  Downloading scikit_tree-0.4.1-cp310-cp310-macosx_10_9_x86_64.whl.metadata (7.3 kB)
Collecting numpy (from scikit-tree)
  Using cached numpy-1.26.2-cp310-cp310-macosx_10_9_x86_64.whl.metadata (61 kB)
Collecting scipy>=1.5.0 (from scikit-tree)
  Using cached scipy-1.11.4-cp310-cp310-macosx_10_9_x86_64.whl.metadata (60 kB)
Collecting scikit-learn>=1.3.1 (from scikit-tree)
  Downloading scikit_learn-1.3.2-cp310-cp310-macosx_10_9_x86_64.whl.metadata (11 kB)
Collecting joblib>=1.1.1 (from scikit-learn>=1.3.1->scikit-tree)
  Using cached joblib-1.3.2-py3-none-any.whl.metadata (5.4 kB)
Collecting threadpoolctl>=2.0.0 (from scikit-learn>=1.3.1->scikit-tree)
  Using cached threadpoolctl-3.2.0-py3-none-any.whl.metadata (10.0 kB)
Downloading scikit_tree-0.4.1-cp310-cp310-macosx_10_9_x86_64.whl (2.1 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 2.3 MB/s eta 0:00:00
Downloading scikit_learn-1.3.2-cp310-cp310-macosx_10_9_x86_64.whl (10.2 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.2/10.2 MB 7.4 MB/s eta 0:00:00
Downloading numpy-1.26.2-cp310-cp310-macosx_10_9_x86_64.whl (20.6 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.6/20.6 MB 7.5 MB/s eta 0:00:00
Downloading scipy-1.11.4-cp310-cp310-macosx_10_9_x86_64.whl (37.3 MB)
   ━━━━━━━━━━━━━━━━━━━━╸━━━━━━━━━━━━━━━━━━━ 19.2/37.3 MB 7.7 MB/s eta 0:00:03
```

I cannot reproduce the issue locally because it seems to work. The [pyproject.toml](https://github.com/neurodata/scikit-tree/blob/main/pyproject.toml) file also seems to be structured correctly. Any idea what might be happening?

**Comment by adam2392** [comment]
Ah I see, yeah we renamed to `scikit-tree` and so the current versions should handle the issue correctly. Sorry, I forgot to mention that.

@adrinjalali

**Comment by adam2392** [comment]
Done! 

Before/after moving the repo, do we need to do any renaming/etc.? 

E.g. where our documentation is hosted is currently https://docs.neurodata.io/scikit-tree/dev/index.html

**Comment by adrinjalali** [comment]
You need to send a transfer request to my personal account, then I will move the repo to the org. You don't need to delete anything, github should handle everything fine. For your local copy, you just change your remote address.

**Comment by adam2392** [comment]
Hi @adrinjalali sorry for the delay. We are figuring out some licensing issues due to Johns Hopkins. Is the following license still permissible for scikit-learn-contrib? 

https://github.com/neurodata/scikit-tree/blob/main/LICENSE

It simply permits anyone to freely use it, but if they make "money off of it", they can't just freely use the software anymore.

**Comment by adrinjalali** [comment]
That makes it in effect a freemium software kinda thing. I'm not sure if that's appropriate for the contrib org.

I'll ping others to see what they think.

**Comment by GaelVaroquaux** [comment]
This license does not seem to be compatible with the standard definitions from OSI (Open Source Initiative). I'm not in favor to have such licenses in scikit-learn-contrib, it would be confusing for users.

**Comment by betatim** [comment]
In https://github.com/scikit-learn-contrib/scikit-learn-contrib/blob/master/workflow.md#requirements there is a statement that an "open-source license" is required. The Polyform license used in scikit-tree seems to be a "bespoke" license, in the sense that it isn't one of the commonly used ones for open-source projects. Reading their website a bit to get a feeling for what Polyform is, who is behind it, etc I found https://polyformproject.org/what-is-polyform/ which states that "Polyform is not Open source or free software.  There are plenty of existing open source licenses. PolyForm is not a substitute for them, but an alternative for those who want to license source code under limited rights."

This kind of fits my personal impression that licenses which "reserve some rights" (for example "for non-commercial use only") are not generally considered "open-source" licenses.

**Comment by glemaitre** [comment]
I'm -1 with the license and I would think that the license should come from: https://en.wikipedia.org/wiki/License_compatibility#Compatibility_of_FOSS_licenses

**Comment by adam2392** [comment]
Sounds good! Thanks for the comments. I will circle back with the admins here!

**Comment by glemaitre** [comment]
By the way, I think that by the terms of the BSD 3-clause license, you cannot redistribute the repo solely with the Polyform License since the source code that comes scikit-learn should keep the original license and copyrights. I might also be wrong because I far to be someone understanding licensing :)
